### PR TITLE
Include mime type params when setting the response content-type

### PR DESCRIPTION
**Pull request by klee97**
Rest.li is setting the response `content-type` based on the best matching `accept` mime type. However it only sets the base mime type while ignoring any mime type parameters. If a mime type parameter is malformed, it is silently ignored.

Changes:
* Include mime type params when setting the response content type
* Throw an InvalidMimeTypeException when there are invalid params in the accept header.

### CODE REVIEW

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 1]
```diff
@@ -67,12 +67,25 @@
     @Override
     public String toString()
     {
-      StringBuffer s = new StringBuffer("('" + type + "', '" + subType
-                                                + "', {");
+      StringBuilder s = new StringBuilder("('").append(type).append("', '").append(subType).append("', {");
       for (String k : params.keySet())
```
Comment (karthikrg) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:71]: Doing a keySet and then get(K) is so wasteful. Please use forEach on the map.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 2]
```diff
@@ -67,12 +67,25 @@
...
       return s.append("})").toString();
     }
+
+    /**
+     * Build the String for the content type. "q" param is ignored
+     */
+    String toContentType()
+    {
+      StringBuilder s = new StringBuilder(type).append("/").append(subType);
+      for (String k : params.keySet()) {
```
Comment (karthikrg) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:82]: Same comment as previous about not using keySet.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 3]
```diff
@@ -67,12 +67,25 @@
...
+        if (!k.equals("q")) {
```
Comment (karthikrg) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:83]: Please move q to a constant and add a comment explaining why it is ignored.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 4]
```diff
@@ -67,12 +67,25 @@
...
       for (String k : params.keySet())
-        s.append("'" + k + "':'" + params.get(k) + "',");
+        s.append("'").append(k).append("':'").append(params.get(k)).append("',");
       return s.append("})").toString();
     }
+
+    /**
+     * Build the String for the content type. "q" param is ignored
+     */
+    String toContentType()
```
Comment (karthikrg) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:79]: This has a lot of duplication with the toString() method. Why not parameterize this method adding a boolean param like excludeQuality and invoke accordingly?

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 5]
```diff
@@ -93,8 +106,12 @@ protected static ParseResults parseMimeType(String mimeType)
     {
       String p = parts[i];
       String[] subParts = StringUtils.split(p, '=');
-      if (subParts.length == 2)
+      if (subParts.length == 2) {
```
Comment (karthikrg) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:109]: Why are we doing this manual parsing here instead of using the mimetype parser?

**Review by karthikrg — CHANGES_REQUESTED** [review]

**Comment by BrianPin** [comment]
Just add a general comment that in pegasus/restli the left bracket is on the next new line.

File: gradle.properties [hunk 6]
```diff
@@ -1,4 +1,4 @@
-version=29.15.4
+version=29.15.5
```
Comment (karthikbalasub) [inline gradle.properties:1]: This was released today. So you'd have to update here and the changelog.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java (hunk 5 above)
Comment (karthikbalasub) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:109]: This MimeParse utility is used to pick the best match based on "q". Not use why we used a custom parser, one difference is that special handling of "*" accept header (lines 118 to 121). This is not handled in jdk MimeType.

**Comment by junchuanwang** [comment]
Is this a backward-compatible change? Cause previously worked header would cause exceptions?

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java (hunk 4 above)
Comment (klee97) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:79]: `toString` is a representation of the ParseResult, while `toContentType` is the actual mime type, so I think it's better to keep them separate. The string format is a little different, so I don't see much that can be generalized into a helper.

**Comment by klee97** [comment]
> Just add a general comment that in pegasus/restli the left bracket is on the next new line.

Thanks @BrianPin, is there a style command I can run?

**Comment by klee97** [comment]
> Is this a backward-compatible change? Cause previously worked header would cause exceptions?

@junchuanwang I suppose if someone is sending incorrect headers, it would start breaking. I can do a major version bump to be safe

File: gradle.properties [hunk 7]
```diff
@@ -1,4 +1,4 @@
-version=29.15.4
+version=30.15.5
```
Comment (karthikbalasub) [inline gradle.properties:1]: This should be 30.0.0

Doing a major version would also mean waiting to merge some other backwards incompatible changes so they can all be bundled in on major version. 
It would add additional time to the release, let us know if that is okay with you.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 8]
```diff
@@ -94,7 +108,14 @@ protected static ParseResults parseMimeType(String mimeType)
...
       String[] subParts = StringUtils.split(p, '=');
       if (subParts.length == 2)
+      {
         results.params.put(subParts[0].trim(), subParts[1].trim());
+      }
+      else
+      {
+        // invalid parameters
+        throw new InvalidMimeTypeException(mimeType);
+      }
```
Comment (karthikbalasub) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:118]: Is this needed? If existing behavior that ignores invalid params is not blocking you, then we can split the change and move the backwards incompatible change to a separate PR.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java [hunk 9]
```diff
@@ -67,12 +69,24 @@
...
+    /**
+     * Build the String for the content type header
+     */
+    String toContentType() {
+      StringBuilder s = new StringBuilder(type).append("/").append(subType);
+      params.forEach((k, v) ->
+      {
+        // Exclude accept type's "q" param from the content type
+        if (!k.equals(QUALITY_PARAM))
+          s.append("; ").append(k).append("=").append(v);
```
Comment (karthikbalasub) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/util/MIMEParse.java:86]: add {}

File: gradle.properties (hunk 7 above)
Comment (klee97) [reply gradle.properties:1]: It's not urgent to fail invalid parameters. We can do as you suggested and split the change.

**Comment by BrianPin** [comment]
> > Just add a general comment that in pegasus/restli the left bracket is on the next new line.
> 
> Thanks @BrianPin, is there a style command I can run?

Nothing I am aware of, I think we probably just need to be manually change this

**Review by karthikbalasub — APPROVED** [review]
